--- numcodecs-0.10.2.orig/setup.py	2022-08-01 16:15:24.000000000 -0300
+++ numcodecs-0.10.2/setup.py	2022-10-31 15:53:07.740308443 -0300
@@ -6,13 +6,7 @@
 from distutils.command.build_ext import build_ext
 from distutils.errors import CCompilerError, DistutilsExecError, \
     DistutilsPlatformError
-
-try:
-    from Cython.Build import cythonize
-except ImportError:
-    have_cython = False
-else:
-    have_cython = True
+from Cython.Build import cythonize
 
 # determine CPU support for SSE2 and AVX2
 cpu_info = cpuinfo.get_cpu_info()
@@ -24,7 +18,7 @@
 
 # setup common compile arguments
 have_cflags = 'CFLAGS' in os.environ
-base_compile_args = list()
+base_compile_args = []
 if have_cflags:
     # respect compiler options set by user
     pass
@@ -104,10 +98,7 @@
     else:
         info('compiling Blosc extension without AVX2 support')
 
-    if have_cython:
-        sources = ['numcodecs/blosc.pyx']
-    else:
-        sources = ['numcodecs/blosc.c']
+    sources = ['numcodecs/blosc.pyx']
 
     # define extension module
     extensions = [
@@ -119,8 +110,7 @@
                   ),
     ]
 
-    if have_cython:
-        extensions = cythonize(extensions)
+    extensions = cythonize(extensions)
 
     return extensions
 
@@ -144,10 +134,7 @@
                      if os.path.isdir(d)]
     # define_macros += [('CYTHON_TRACE', '1')]
 
-    if have_cython:
-        sources = ['numcodecs/zstd.pyx']
-    else:
-        sources = ['numcodecs/zstd.c']
+    sources = ['numcodecs/zstd.pyx']
 
     # define extension module
     extensions = [
@@ -159,8 +146,7 @@
                   ),
     ]
 
-    if have_cython:
-        extensions = cythonize(extensions)
+    extensions = cythonize(extensions)
 
     return extensions
 
@@ -177,10 +163,7 @@
     include_dirs += ['numcodecs']
     # define_macros += [('CYTHON_TRACE', '1')]
 
-    if have_cython:
-        sources = ['numcodecs/lz4.pyx']
-    else:
-        sources = ['numcodecs/lz4.c']
+    sources = ['numcodecs/lz4.pyx']
 
     # define extension module
     extensions = [
@@ -192,8 +175,7 @@
                   ),
     ]
 
-    if have_cython:
-        extensions = cythonize(extensions)
+    extensions = cythonize(extensions)
 
     return extensions
 
@@ -208,10 +190,7 @@
     include_dirs = ['numcodecs']
     # define_macros += [('CYTHON_TRACE', '1')]
 
-    if have_cython:
-        sources = ['numcodecs/vlen.pyx']
-    else:
-        sources = ['numcodecs/vlen.c']
+    sources = ['numcodecs/vlen.pyx']
 
     # define extension module
     extensions = [
@@ -223,8 +202,7 @@
                   ),
     ]
 
-    if have_cython:
-        extensions = cythonize(extensions)
+    extensions = cythonize(extensions)
 
     return extensions
 
@@ -234,10 +212,7 @@
 
     extra_compile_args = list(base_compile_args)
 
-    if have_cython:
-        sources = ['numcodecs/compat_ext.pyx']
-    else:
-        sources = ['numcodecs/compat_ext.c']
+    sources = ['numcodecs/compat_ext.pyx']
 
     # define extension module
     extensions = [
@@ -246,8 +221,7 @@
                   extra_compile_args=extra_compile_args),
     ]
 
-    if have_cython:
-        extensions = cythonize(extensions)
+    extensions = cythonize(extensions)
 
     return extensions
 
@@ -257,10 +231,7 @@
 
     extra_compile_args = list(base_compile_args)
 
-    if have_cython:
-        sources = ['numcodecs/_shuffle.pyx']
-    else:
-        sources = ['numcodecs/_shuffle.c']
+    sources = ['numcodecs/_shuffle.pyx']
 
     # define extension module
     extensions = [
@@ -269,8 +240,7 @@
                   extra_compile_args=extra_compile_args),
     ]
 
-    if have_cython:
-        extensions = cythonize(extensions)
+    extensions = cythonize(extensions)
 
     return extensions
 
@@ -321,7 +291,7 @@
         cmdclass = dict(build_ext=ve_build_ext)
     else:
         ext_modules = []
-        cmdclass = dict()
+        cmdclass = {}
 
     setup(
         name='numcodecs',
@@ -334,7 +304,8 @@
         },
         setup_requires=[
             'setuptools>18.0',
-            'setuptools-scm>1.5.4'
+            'setuptools-scm>1.5.4',
+            'Cython',
         ],
         install_requires=[
             'entrypoints',
@@ -363,6 +334,7 @@
             "Programming Language :: Python :: 3.8",
             "Programming Language :: Python :: 3.9",
             "Programming Language :: Python :: 3.10",
+            "Programming Language :: Python :: 3.11",
         ],
         author='Alistair Miles',
         author_email='alimanfoo@googlemail.com',
